### PR TITLE
refactor

### PR DESCRIPTION
**Pull request by jonschlinkert**
@doowb please review and give feedback if you see anything that needs to be fixed/improved. Otherwise feel free to merge.

### CODE REVIEW

File: lib/async.js [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+  if (Array.isArray(basedir)) {
+    return readdirs(basedir, options);
+  }
+
+  const opts = { ...defaults, ...options };
+
+  const isMaxDepth = file => file.depth >= opts.depth;
+  const isIgnoredFile = opts.isIgnoredFile || noop(false);
+  const isIgnoredDir = opts.isIgnoredDir || noop(false);
+  const isMatch = opts.isMatch ? utils.matcher(opts.isMatch, opts) : () => true;
```
Comment (doowb) [inline lib/async.js]: Suggestion: use `noop(true)` here for consistency:

```suggestion
  const isMatch = opts.isMatch ? utils.matcher(opts.isMatch, opts) : noop(true);
```

File: lib/async.js [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+  const recurse = opts.recursive === true || (typeof options.depth === 'number' && options.depth > 1);
+
+  const getReturnValue = file => {
+    if (file.keep === false) return;
+    if (file.ignore === true) return;
+
+    if (file.keep !== true) {
+      if (file.isSymbolicLink() && symlinks !== true) return;
+      if (file.isDirectory() && options.nodir === true) return;
+      if (options.dot === false && file.name[0] !== '.') return;
```
Comment (doowb) [inline lib/async.js]: Is this supposed to be `file.name[0] === '.'`?

```suggestion
      if (options.dot === false && file.name[0] === '.') return;
```

File: lib/async.js [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+          } catch (err) {
+            file[kRealPathExists] = false;
+
+            if (err.code !== 'ENOENT') {
+              next(err);
+              return;
+            }
+          }
+        } else if (opts.stat === true) {
+          file.stat = await fs.promises.stat(file[kPath]);
```
Comment (doowb) [inline lib/async.js]: Just curious... any reason to not add `file.isFile()` and `file.isDirectory()` here?

File: lib/async.js [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+
+        if (!file.relative) {
+          if (file[kPath] === file.path && file.base === file.cwd) {
+            file.relative = path.join(dirent.relative, file.name);
+          } else {
+            file.relative = path.relative(file.base, file.path);
+          }
+        }
+
+        if (file.isDirectory()) {
```
Comment (doowb) [inline lib/async.js:204]: Based on my comment above... I'm not assuming that a `Dirent` already has `isFile()` and `isDirectory()`, so when handling the symlinks, those are added to override the default functionality.

File: lib/sync.js [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+  if (Array.isArray(basedir)) {
+    return readdirsSync(basedir, options);
+  }
+
+  const opts = { ...defaults, ...options };
+
+  const isMaxDepth = file => file.depth >= opts.depth;
+  const isIgnoredFile = opts.isIgnoredFile || noop(false);
+  const isIgnoredDir = opts.isIgnoredDir || noop(false);
+  const isMatch = opts.isMatch ? utils.matcher(opts.isMatch, opts) : () => true;
```
Comment (doowb) [inline lib/sync.js]: Same suggestion as in `async`:

```suggestion
  const isMatch = opts.isMatch ? utils.matcher(opts.isMatch, opts) : noop(true);
```

File: lib/sync.js [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+  const recurse = opts.recursive === true || (typeof options.depth === 'number' && options.depth > 1);
+
+  const getReturnValue = file => {
+    if (file.keep === false) return;
+    if (file.ignore === true) return;
+
+    if (file.keep !== true) {
+      if (file.isSymbolicLink() && symlinks !== true) return;
+      if (file.isDirectory() && options.nodir === true) return;
+      if (options.dot === false && file.name[0] !== '.') return;
```
Comment (doowb) [inline lib/sync.js]: Same question above...

```suggestion
      if (options.dot === false && file.name[0] === '.') return;
```

File: lib/async.js [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+  });
+};
+
+const readdirs = (dirs, options = {}) => {
+  const unique = options.unique === true;
+  const files = unique ? new Set() : [];
+  const pending = [];
+
+  const onPush = file => {
+    files[unique ? 'add' : 'push'](file[kResult]);
```
Comment (doowb) [inline lib/async.js]: In `sync`, `options.onPush` is called here if it exists... should that be done here too?

File: package.json [hunk 8]
```diff
@@ -1,9 +1,12 @@
 {
-  "name": "@folder/readdir",
+  "name": "readdir",
```
Comment (doowb) [inline package.json]: Was this supposed to change?

```suggestion
  "name": "@folder/readdir",
```

**Review by doowb — CHANGES_REQUESTED** [review]
Looks awesome!
I left a couple of comments/suggestions inline, but I don't see anything that really sticks out.

**Comment by doowb** [comment]
@jonschlinkert I left some review comments.

File: lib/async.js (hunk 3 above)
Comment (jonschlinkert) [reply lib/async.js]: Yes, since it's not a symbolic link, those functions will already return the correct result.

File: lib/async.js (hunk 4 above)
Comment (jonschlinkert) [reply lib/async.js:204]: Correct. No overrides are needed when the file is not a symbolic link.

File: lib/async.js (hunk 7 above)
Comment (jonschlinkert) [reply lib/async.js]: Good catch.

File: package.json (hunk 8 above)
Comment (jonschlinkert) [reply package.json]: Good catch. Not sure how this got changed.

File: test/readdir.js [hunk 9]
```diff
@@ -7,66 +7,70 @@ const assert = require('assert');
...
-  process.on('exit', () => deleteFixtures());
+  after(() => cleanupTemp());
 
   describe('no options', () => {
     it('should read files in a directory and return a promise with files', () => {
       return readdir(__dirname)
         .then(files => {
           assert(files.some(file => path.basename(file) === 'readdir.js'));
-          assert(files.some(file => path.basename(file) === 'fixtures'));
+          assert(files.some(file => path.basename(file) === 'temp'));
```
Comment (doowb) [inline test/readdir.js]: Looks like this test is failing because `temp` doesn't exist at this point (it's deleted in `beforeEach`).

Comment (doowb) [reply test/readdir.js]: Oh... I think it's that `temp` just doesn't exist yet in a newly cloned repo. (I just noticed that `temp` isn't deleted, just the files inside)

**Pull request merged**